### ISSUE TITLE
膨胀卷积后的mask数据集有吗

### ISSUE THREAD

**Issue by laoxie521** [jinwan1994/SG-ShadowNet#3]

**Comment by laoxie521** [comment]
请问膨胀卷积后的mask数据集有吗，还有我发现batchsize只要大于1，train出来的图片就会失真，我查看发现tensor2img这个方法有一部压缩操作，似乎只能调整batchsize为1。第二那个mask_50我直接把阴影mask通过膨胀函数扩大50倍，结果训练的图像整体都很模糊，不知道是不是这一步出了问题，望能解答

**Comment by HarryHsing** [comment]
老哥有了吗？

**Comment by laoxie521** [comment]
> 老哥有了吗？

这个需要自己做，用cv方法把已有mask数据集进行膨胀，用文件夹存起来

**Comment by HarryHsing** [comment]
> > 老哥有了吗？
> 
> 这个需要自己做，用cv方法把已有mask数据集进行膨胀，用文件夹存起来

ok, 多谢老哥

**Comment by ZiyangBo** [comment]
import torch
import torch.nn.functional as F
import os
import cv2
import numpy as np


def dilate_mask_torch(mask, kernel_size=50):
    mask_tensor = torch.tensor(mask, dtype=torch.float32).unsqueeze(0).unsqueeze(0).cuda()  

    dilated_mask = F.max_pool2d(mask_tensor, kernel_size, stride=1, padding=kernel_size // 2)

    dilated_mask = (dilated_mask > 0).float().squeeze().cpu().numpy() * 255  
    dilated_mask = dilated_mask.astype(np.uint8)

    dilated_mask = cv2.resize(dilated_mask, (640, 480), interpolation=cv2.INTER_NEAREST)

    return dilated_mask


def batch_process_masks_torch(input_folder, output_folder, kernel_size=50):
    if not os.path.exists(output_folder):
        os.makedirs(output_folder)

    mask_files = [f for f in os.listdir(input_folder) if f.endswith('.png')]

    for filename in mask_files:
        input_path = os.path.join(input_folder, filename)
        output_path = os.path.join(output_folder, filename)

        mask = cv2.imread(input_path, cv2.IMREAD_GRAYSCALE)
        if mask is None:
            print(f"警告: 无法读取 {input_path}，跳过该文件。")
            continue

        mask[mask > 0] = 255
        dilated_mask = dilate_mask_torch(mask, kernel_size)

        cv2.imwrite(output_path, dilated_mask)
        print(f"处理完成: {input_path} → {output_path}")


# 批量处理
input = 'dataset/AISTD/train/train_B'
output = 'AISTD/train/train_mask50'
batch_process_masks_torch(input, output, kernel_size=50)

**Comment by ZiyangBo** [comment]
`import torch
import torch.nn.functional as F
import os
import cv2
import numpy as np

def dilate_mask_torch(mask, kernel_size=50):

mask_tensor = torch.tensor(mask, dtype=torch.float32).unsqueeze(0).unsqueeze(0).cuda()  
dilated_mask = F.max_pool2d(mask_tensor, kernel_size, stride=1, padding=kernel_size // 2)

dilated_mask = (dilated_mask > 0).float().squeeze().cpu().numpy() * 255 
dilated_mask = dilated_mask.astype(np.uint8)

dilated_mask = cv2.resize(dilated_mask, (640, 480), interpolation=cv2.INTER_NEAREST)

return dilated_mask
def batch_process_masks_torch(input_folder, output_folder, kernel_size=50):
if not os.path.exists(output_folder):
os.makedirs(output_folder)

mask_files = [f for f in os.listdir(input_folder) if f.endswith('.png')]

for filename in mask_files:
    input_path = os.path.join(input_folder, filename)
    output_path = os.path.join(output_folder, filename)

    mask = cv2.imread(input_path, cv2.IMREAD_GRAYSCALE)
    if mask is None:
        print(f"警告: 无法读取 {input_path}，跳过该文件。")
        continue

    mask[mask > 0] = 255
    dilated_mask = dilate_mask_torch(mask, kernel_size)

    cv2.imwrite(output_path, dilated_mask)
    print(f"处理完成: {input_path} → {output_path}")
批量处理
input = 'dataset/AISTD/train/train_B'
output = 'dataset/AISTD/train/train_mask50'
batch_process_masks_torch(input, output, kernel_size=50)

`
